typing/ctype.ml

```diff
@@ -689,11 +689,19 @@ let check_scope_escape level ty =
       );
       begin match ty.desc with
       | Tconstr (p, _, _) when level < Path.scope p ->
-        let var = newvar2 level in
-        raise (Unify [(ty,ty); (var, var)])
-      | _ -> ()
+          begin match !forward_try_expand_once env ty with
+          | ty' -> aux ty'
+          | exception Cannot_expand ->
+              let var = newvar2 level in
+              raise (Unify [(ty,ty); (var, var)])
+          end
+      | Tpackage (p, _, _) when level < Path.scope p ->
+          let p' = normalize_package_path env p in
+          if Path.same p p' then raise (Unify [(ty, newvar2 level)]);
+          iter_type_expr aux ty
```

Comment:
I first thought that something was fishy because here because you are iterating on `ty`, and I would expect that `Tpackage(p', ...)` should be used instead. But actually I guess that doesn't matter as `iter_type_expr` only recurses on the child, which are the same. Right? (It should probably be called `iter_on_type_children` or something.)

Reply:
> iter_type_expr only recurses on the child, which are the same. Right?

Right.

> p' could in theory also be below the scope. Could you rebuild ty' and call aux again?

Nice catch.
I amended the commit once more.

Thank you for all the hand holding.